pkg/apis/networking/v1/conversion.go

```diff
@@ -0,0 +1,138 @@
+/*
+ Copyright 2021 The Hybridnet Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+
+	"github.com/alibaba/hybridnet/pkg/constants"
+	"github.com/alibaba/hybridnet/pkg/ipam/strategy"
+)
+
+const (
+	IPInstanceV11           = "v1.1"
+	IPInstanceV12           = "v1.2"
+	IPInstanceLatestVersion = IPInstanceV12
+)
+
+func parseIPInstanceVersion(ipInstance *IPInstance) string {
+	if len(ipInstance.Labels[constants.LabelVersion]) == 0 {
+		return IPInstanceV11
+	}
+	return ipInstance.Labels[constants.LabelVersion]
+}
+
+func CanonicalizeIPInstance(cluster cluster.Cluster) (err error) {
+	var c client.Client
+	if c, err = client.New(cluster.GetConfig(), client.Options{
+		Scheme: cluster.GetScheme(),
+		Mapper: cluster.GetRESTMapper(),
+	}); err != nil {
+		return err
+	}
+
+	getPodUID := func(namespace, name string) (types.UID, error) {
+		pod := &corev1.Pod{}
+		if err := c.Get(context.TODO(),
+			types.NamespacedName{
+				Namespace: namespace,
+				Name:      name,
+			},
+			pod); err != nil {
+			if err = client.IgnoreNotFound(err); err == nil {
+				return "", nil
+			}
+			return "", err
+		}
+		return pod.UID, nil
+	}
+
+	var ipInstanceList = &IPInstanceList{}
+	if err = c.List(context.TODO(), ipInstanceList); err != nil {
+		return err
+	}
+
+	for i := range ipInstanceList.Items {
+		var ipInstance = &ipInstanceList.Items[i]
+		if parseIPInstanceVersion(ipInstance) == IPInstanceLatestVersion {
+			continue
+		}
+
+		if err = convertIPInstanceToLatestVersion(ipInstance, getPodUID); err != nil {
```

Comment:
Has the conversion be updated to apiserver?

Reply:
fixed

Reply:
fixed